### DISCUSSION TITLE
(Question) Python | Merging multiple JDM jsons together for common input

### DISCUSSION THREAD

**Discussion by tushar-sahay-np** [gorules/zen#63] [Q&A]
Hello, 

I have a use case where we have multiple json files (in different s3 folders) which all have the same input (a Transaction object). I want to run all the rules in the multiple json files against a common input i.e transaction object. 
For e.g: I have 3 different json files : rule1.json, rul2.json, and rule3.json. They contain 1 decision table each. They all work on the same input i.e transaction, and post rule evaluation, the output should be a merged output of all the decision tables together. 

How do I acheive this in python? 

One way is I can serially evaluate the rules from the JDM (i.e process one by one and collate the output at the end). 

But I was wondering if there a more suitable way to achieve this?

**Comment by ivanmiletic** [comment]
Hi @tushar-sahay-np yes there is, you would be using loaders and one parent JDM file that has references to these 3. For example, besides rule1.json, rule2.json and rule3.json you would have rule_aggr.json. Free editor doesn't have this functionality exposed but our cloud solution does and soon we will have self hosted brms. You can try it there and create one graph that has references to children.

<img width="1177" alt="Screenshot 2023-07-27 at 16 00 53" src="https://github.com/gorules/zen/assets/60513195/37a7bc1b-fdbd-4dd7-8466-3bbc98649668">

Please note for this to work you will need to setup loader properly to load based on a "key".
Example JSON:

`{
  "contentType": "application/vnd.gorules.decision",
  "nodes": [
    {
      "id": "900ae694-2ccf-42e5-be84-c714dc258d53",
      "type": "inputNode",
      "position": {
        "x": 90,
        "y": 230
      },
      "name": "Request"
    },
    {
      "id": "b90b0e01-0e91-42d1-b832-41b84830750b",
      "type": "decisionNode",
      "position": {
        "x": 370,
        "y": 100
      },
      "name": "Rule 1",
      "content": {
        "key": "rule1.json"
      }
    },
    {
      "id": "7b2793a1-a168-48d4-a390-0fc20bdca2a2",
      "type": "decisionNode",
      "position": {
        "x": 370,
        "y": 230
      },
      "name": "Rule 2",
      "content": {
        "key": "rule2.json"
      }
    },
    {
      "id": "1e6b2def-08f1-4801-a2de-03154463098a",
      "type": "decisionNode",
      "position": {
        "x": 370,
        "y": 360
      },
      "name": "Rule 3",
      "content": {
        "key": "rule3.json"
      }
    },
    {
      "id": "f445e8f7-56df-4ec6-a5ec-10a8f3b159f3",
      "type": "outputNode",
      "position": {
        "x": 660,
        "y": 230
      },
      "name": "Response"
    }
  ],
  "edges": [
    {
      "id": "eeb318ce-f4e2-4ede-8762-bc35fed6f2a6",
      "sourceId": "900ae694-2ccf-42e5-be84-c714dc258d53",
      "type": "edge",
      "targetId": "b90b0e01-0e91-42d1-b832-41b84830750b"
    },
    {
      "id": "1c382fd1-8d1b-42d6-b460-9a236ebbfc05",
      "sourceId": "900ae694-2ccf-42e5-be84-c714dc258d53",
      "type": "edge",
      "targetId": "7b2793a1-a168-48d4-a390-0fc20bdca2a2"
    },
    {
      "id": "f0c31549-3640-437c-99b5-449339637d27",
      "sourceId": "900ae694-2ccf-42e5-be84-c714dc258d53",
      "type": "edge",
      "targetId": "1e6b2def-08f1-4801-a2de-03154463098a"
    },
    {
      "id": "0b288852-1054-44f8-940e-2b3e6bc06cf3",
      "sourceId": "b90b0e01-0e91-42d1-b832-41b84830750b",
      "type": "edge",
      "targetId": "f445e8f7-56df-4ec6-a5ec-10a8f3b159f3"
    },
    {
      "id": "1e6828b2-4f10-4ec2-9130-d02125623a4d",
      "sourceId": "7b2793a1-a168-48d4-a390-0fc20bdca2a2",
      "type": "edge",
      "targetId": "f445e8f7-56df-4ec6-a5ec-10a8f3b159f3"
    },
    {
      "id": "e351be66-7171-4b54-ad08-91e5aaa6d3a3",
      "sourceId": "1e6b2def-08f1-4801-a2de-03154463098a",
      "type": "edge",
      "targetId": "f445e8f7-56df-4ec6-a5ec-10a8f3b159f3"
    }
  ]
}`

**Reply by tushar-sahay-np** [reply]
Thanks for the quick response @ivanmiletic. 

So in the above example, what will be the 'key' passed to the loader if I want the rules to run on all rule files together (rule1.json, rule2.json, and rule3.json)?

Would it be something like - 

```
import zen

def loader(key):
    with open("./jdm_directory/" + key, "r") as f:
        return f.read()

engine = zen.ZenEngine({"loader": loader})
result = engine.evaluate("rule_aggr.json", {"input": 5})
```

**Reply by ivanmiletic** [reply]
Key is the filename/path that you have referenced in your parent graph and it is usually a path to your sub decision. Imagine you have 4 decision files and they are named rule_aggr.json, rule1.json, rule2.json and rule3.json, your parent references them, when engine evaluates it passes the key to your loader so you can "load" file content and pass it to engine for evaluation.

In this case we have file system example but you can also create S3 loader. You get a key and based on the key you read a content of file from s3 and pass it to engine for evaluation by returning value.
